File: src/main/scala/require4s/Module.scala
package require4s

/**
 * Created by sxend on 14/04/28.
 */
trait Module[M] {
  self =>
  def export: M

  val id: String = self.getClass.getName
}

object Module {
  def wrap[M](module: Module[M], f: => M): Module[M] = new Module[M] {
    override val id = module.id

    override def export: M = f
  }
}

File: src/main/java/require4s/Module.java
package require4s;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Created by sxend on 2014/05/02.
 */
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
public @interface Module {
    Class<?> value();
}
